feat(target): add support for target.page for 'backgroud_page'

yanivefraim opened the pull request:
Fixes #2438

yanivefraim committed:
feat(target): add support for target.page for backgroud_page

aslushnikov commented on lib/Target.js:
```diff
@@ -38,7 +38,7 @@ class Target {
    */
   async page() {
-    if (this._targetInfo.type === 'page' && !this._pagePromise) {
+    if ((this._targetInfo.type === 'page' || this._targetInfo.type === 'background_page') && !this._pagePromise) {
```
can you please update `target.page()` documentation and mention that it now works with `'background_page'` type?

Also: can you please add the `'background_page'` type to  `target.type()` documentation?

yanivefraim replied:
`target.page()` docs were updated. 

Regarding `target.type()` - it was already there :)

aslushnikov commented on test/target.spec.js:
```diff
@@ -45,6 +45,15 @@ module.exports.addTests = function({testRunner, expect, puppeteer, browserWithEx
+      const backgroundPageTarget = targets.find(target => target.type() === 'background_page');
+      const page = await backgroundPageTarget.page();
+      expect(page).not.toBeNull();
+      await page.close();
```
Can you please try evaluating something inside the page? It should be working, but I'd like to be sure.

Also: it's a little crazy to do `page.close` - iirc background page lifetime is controlled automatically. Let's drop the line?

yanivefraim replied:
done. 10x!

yanivefraim committed:
Update api.md

yanivefraim committed:
verify page is working using evaluate

yanivefraim commented:
@aslushnikov - after rebasing master I have the [following test](https://github.com/GoogleChrome/puppeteer/blob/master/test/puppeteer.spec.js#L273) failing (pushed it red to my branch intentionally, so you could see the error). 

The reason is that we now have a background page as well as the default `[about:blank]`, which I just added. 

I am wondering if this is an indicator that we just broke backward compatibility here (by automatically adding something like `chrome-extension://kmendfapggjehodndflmmgagdbamhnfd/_generated_background_page.html` page to everyone)?

aslushnikov commented:
@yanivefraim good point.

It looks like the right thing to do here would be to have `target.page()` working for both background pages and regular pages, whereas `browser.pages()` should return only regular pages.

The motivation behind this is that `browser.pages()` method is perceived as browser visible tabs.

yanivefraim commented:
> It looks like the right thing to do here would be to have target.page() working for both background pages and regular pages, whereas browser.pages() should return only regular pages.

Done, the test is now green again :)
10x!

yanivefraim committed:
add missing dependency after rebase

yanivefraim committed:
move test to headful.spec.js

yanivefraim committed:
remove redundant tests

yanivefraim commented:
Something went wrong here, after rebasing (had conflicts and now the test we discussed [here](https://github.com/GoogleChrome/puppeteer/pull/2600#issuecomment-392290756) is failing again). I will take a look later today and fix it.

yanivefraim committed:
browser.pages will not list background pages

yanivefraim commented:
> Something went wrong here, after rebasing (had conflicts and now the test we discussed here is failing again). I will take a look later today and fix it.

For some reason rebase removed my last fix (no conflicts, just removed my code)... 

Added back the fix for this:
>It looks like the right thing to do here would be to have target.page() working for both background pages and regular pages, whereas browser.pages() should return only regular pages.

@aslushnikov - I believe the PR is ready again :)

JoelEinbinder commented on docs/api.md:
```diff
@@ -523,7 +523,7 @@ Disconnects Puppeteer from the browser, but leaves the Chromium process running.
 
 #### browser.pages()
-- returns: <[Promise]<[Array]<[Page]>>> Promise which resolves to an array of all open pages.
+- returns: <[Promise]<[Array]<[Page]>>> Promise which resolves to an array of all open pages. Non visible pages, such as `"background_page"` and `"service_worker"`, will not be listed here. You can find them using [target.page()](#targetpage).
```
nit: service_worker isn't a page, so I think it isn't worth being mentioned here.

yanivefraim replied:
done. 10x

JoelEinbinder commented on lib/Browser.js:
```diff
@@ -183,7 +183,12 @@ class Browser extends EventEmitter {
    */
   async pages() {
-    const pages = await Promise.all(this.targets().map(target => target.page()));
+    const nonVisiblePageTypes = ['service_worker', 'background_page'];
```
I think this can be done simpler as something like
```js
return Promise.all(this.targets().filter(target => target.type === 'page').map(target => target.page())));
```

yanivefraim replied:
@JoelEinbinder - done, 10x!

yanivefraim committed:
refactor

yanivefraim committed:
remove ide stuff
